src/common/storage/src/logging_layer.rs

```diff
@@ -0,0 +1,754 @@
+// Copyright 2023-2025 The Apache Software Foundation
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This layer is originally from OpenDAL's LoggingLayer.
+/// OpenDAL's LoggingLayer logs detailed information at DEBUG level, which is very useful
+/// for debugging storage related issues. But the cost (even `to_string` is expensive)
+/// cannot be ignored with high throughput, which cannot be prevented even logging level is
+/// higher than DEBUG.
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use log::log;
+use log::log_enabled;
+use log::Level;
+use opendal::raw::oio;
+use opendal::raw::Access;
+use opendal::raw::AccessorInfo;
+use opendal::raw::Layer;
+use opendal::raw::LayeredAccess;
+use opendal::raw::OpCopy;
+use opendal::raw::OpCreateDir;
+use opendal::raw::OpDelete;
+use opendal::raw::OpList;
+use opendal::raw::OpPresign;
+use opendal::raw::OpRead;
+use opendal::raw::OpRename;
+use opendal::raw::OpStat;
+use opendal::raw::OpWrite;
+use opendal::raw::Operation;
+use opendal::raw::RpCopy;
+use opendal::raw::RpCreateDir;
+use opendal::raw::RpDelete;
+use opendal::raw::RpList;
+use opendal::raw::RpPresign;
+use opendal::raw::RpRead;
+use opendal::raw::RpRename;
+use opendal::raw::RpStat;
+use opendal::raw::RpWrite;
+use opendal::Buffer;
+use opendal::Error;
+use opendal::ErrorKind;
+use opendal::Metadata;
+
+struct LoggingContext<'a>(&'a [(&'a dyn Display, &'a dyn Display)]);
+
+static LOGGING_TARGET: &str = "opendal::services";
+
+#[derive(Debug, Copy, Clone, Default)]
+struct Logger;
+
+impl Display for LoggingContext<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (k, v) in self.0.iter() {
+            write!(f, " {k}={v}")?;
+        }
+        Ok(())
+    }
+}
+
+impl Logger {
+    #[inline]
+    fn debug_enabled(&self) -> bool {
+        log_enabled!(target: LOGGING_TARGET, Level::Debug)
```

Comment:
This cannot work because we have a lot of logging dispatchers. If one of them on Debug level, this macro will always return true.

Reply:
try solve this in https://github.com/databendlabs/databend/pull/19077/commits/93ea2ab59c0f2cea71bef6a1a219384893453906